### PR TITLE
test scripts DB model, seeding script and DB schema changes, API calls

### PR DESCRIPTION
**Pull request by bshresthap**

### CODE REVIEW

File: Tests/reviews.model.test.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+const ReviewsModel = Reviews.ReviewsModel;
+describe("Reviews model test", () => {
+  beforeAll(async () => {
+    await ReviewsModel.remove({}); //remove all the documents from the collection before everything
+  });
+
+  afterEach(async () => {
+    await ReviewsModel.remove({}); //remove all the document from the collection after each test
+  });
```
Comment (aleksebastian) [inline Tests/reviews.model.test.js:15]: I believe you can use the mocha `beforeEach()` hook instead of having the `beforeAll()` and `afterEach()` hooks. `beforeEach()` runs before each test in the block.

File: Tests/reviews.model.test.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+          batteryLife: 4
+        });
+
+      await mockReview.save();
+      mockReview.configuration = 'echo voice configuration';
+      const updatedReview = await mockReview.save();
+
+      const expected = 'echo voice configuration';
+      const actual = updatedReview.configuration;
+      expect(actual).toEqual(expected);
```
Comment (aleksebastian) [inline Tests/reviews.model.test.js:99]: As I understand this, you're checking if a product information is being updated by saving the product info object with the same product id to check if the product information was updated in the database. If so, by clearing the database after each test with the mocha `afterEach()` hook declared on line 13 the record would just be saved again for the first time instead of being updated.

File: Tests/reviews.service.test.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      //     })
+      //   })
+      //   .then(insertedReview => {
+      //     console.log('insertedReview: ', insertedReview);
+      //     const actual = insertedReview.configuration;
+      //     console.log('actual: ', actual);
+      //     const expected = 'Alexa auto sense temp 1004';
+      //     expect(actual).toEqual(expected);
+      //   })
+      //   .catch(err => console.log(err));
```
Comment (aleksebastian) [inline Tests/reviews.service.test.js:84]: Consider deleting unused code or adding a comment pointing out why the commented code has been left there.

File: database-mongoose/database.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+//     });
+
+//   });
+
+
+// }
+
+// module.exports.insertSeedData = insertSeedData;
+//module.exports.ReviewsModel = ReviewsModel;
```
Comment (aleksebastian) [inline database-mongoose/database.js]: I see you separated your models into different files. Consider deleting the unused code in this file.

File: database-mongoose/reviews.service.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
+const Reviews = require("../database-mongoose/reviews.model");
+let dbModel = require('./database.js');
+const ReviewsModel = Reviews.ReviewsModel;
+
+let insertSeedData = (arrayOfObjects) => {
+
+  //delete model to remove data
+  ReviewsModel.remove({}, function (err) {
+    console.log('collection removed')
+  });
```
Comment (aleksebastian) [inline database-mongoose/reviews.service.js]: You might want to add an if statement in case of error 
`if (err) {
log error
} else {
 do something }`

File: database-mongoose/reviews.service.js [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+  // ReviewsModel.insertMany(arrayOfObjects)
+  //   .then(function () =>
+  //     console.log("Data inserted")
+  //   )
+  //   .catch(err =>
+  //     console.log(err);
+  //   })
```
Comment (aleksebastian) [inline database-mongoose/reviews.service.js]: Unused code

File: database-mongoose/reviews.service.js [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        qualityOfMaterial: element.qualityOfMaterial,
+        batteryLife: element.batteryLife
+
+      }
+    );
+    reviewDocument.save(function (err, doc) {
+      if (err) {
+        console.log(err);
+        return err;
+      }
```
Comment (aleksebastian) [inline database-mongoose/reviews.service.js]: You might want to use an else statement here for clarity purposes

File: database-mongoose/reviews.service.js [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+            {
+              _id: '$productId',
+              avgRating: { $avg: "$rating" },
+              totalRatings: { $sum: 1 }
+            }
+          }
+        ]
+      )
+
+    })
```
Comment (aleksebastian) [inline database-mongoose/reviews.service.js:85]: This looks good!

File: database-mongoose/reviews.service.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        fiveStar: (countByReviews.find(obj => obj._id === 5) ? ((countByReviews.find(obj => obj._id === 5).reviewCount) / ratingsByCustomer[0].totalRatings * 100).toFixed(2).concat('%') : '0%'),
+        fourStar: (countByReviews.find(obj => obj._id === 4) ? ((countByReviews.find(obj => obj._id === 4).reviewCount) / ratingsByCustomer[0].totalRatings * 100).toFixed(2).concat('%') : '0%'),
+        threeStar: (countByReviews.find(obj => obj._id === 3) ? ((countByReviews.find(obj => obj._id === 3).reviewCount) / ratingsByCustomer[0].totalRatings * 100).toFixed(2).concat('%') : '0%'),
+        twoStar: (countByReviews.find(obj => obj._id === 2) ? ((countByReviews.find(obj => obj._id === 2).reviewCount) / ratingsByCustomer[0].totalRatings * 100).toFixed(2).concat('%') : '0%'),
+        oneStar: (countByReviews.find(obj => obj._id === 1) ? ((countByReviews.find(obj => obj._id === 1).reviewCount) / ratingsByCustomer[0].totalRatings * 100).toFixed(2).concat('%') : '0%')
+      };
+      console.log(reviewSummary);
+      var promiseSummary = new Promise((resolve, reject) => {
+        resolve(reviewSummary)
+      });
```
Comment (aleksebastian) [inline database-mongoose/reviews.service.js:101]: You might want to handle your reject case here

File: database-mongoose/reviews.service.js [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        techSupport: result[0].techSupport.toFixed(1),
+        valueForMoney: result[0].valueForMoney.toFixed(1),
+        qualityOfMaterial: result[0].qualityOfMaterial.toFixed(1),
+        batteryLife: result[0].batteryLife.toFixed(1)
+
+      };
+
+      var reviewfeaturesPromise = new Promise((resolve, reject) => {
+        resolve(reviewsByFeature)
+      })
```
Comment (aleksebastian) [inline database-mongoose/reviews.service.js:142]: Likewise here with the reject case.

File: database-mongoose/seeder.js [hunk 11]
```diff
@@ -1,85 +1,86 @@
...
   }
+  return allTestData;
+
 }
-//console.log('allTestData : ', allTestData)
-dbModel.insertSeedData(allTestData);
+
+var testData = generateSeedData(11, 100);//11 customers, 100 products
+//console.log(testData);
+ReviewService.insertSeedData(testData);
```
Comment (aleksebastian) [inline database-mongoose/seeder.js:80]: This looks good. Looks intricate but looks like you made it work! Something minor I noticed was some left over commented out console.logs

File: server/server.js [hunk 12]
```diff
@@ -1,12 +1,57 @@
...
+    })
+    .catch(err => console.log('err: ', err));
+});
+
+app.get('/Reviews/getReviewExcerpts/:productId', (req, res) => {
+  let prodId = req.params; //{id: "5"}
+  db.getReviewExcerpts(parseInt(prodId.productId))
+    .then(results => res.send(results))
+    .catch(err => console.log('err: ', err));
+});
```
Comment (aleksebastian) [inline server/server.js:54]: This looks good! One minor suggestion would be to set up the server so that it handles cases were an invalid id is requested or one that does not exist. I'd also add status codes along with your responses i.e. `res.status(...).send(...)`

**Review by aleksebastian — COMMENTED** [review]
Good job! I can tell you went through a great deal of writing code to get this set up going. Overall, some minor things I saw recurrently was a lot of commented out code not a broad coverage of edge cases. But overall looks good!

File: Tests/reviews.model.test.js (hunk 1 above)
Comment (bshresthap) [reply Tests/reviews.model.test.js:15]: I am using Jest. I agree beforeEach will make more sense. Made a note of it and will use for next PR. Thanks!

File: Tests/reviews.model.test.js (hunk 2 above)
Comment (bshresthap) [reply Tests/reviews.model.test.js:99]: In update test, I have another record in line 66 I am saving and not using the one from insert line 27. Line 93 is saving the new record and line 95 is updating the record and both of these are happening in same test and my understanding is that after each with run after this test completes and not after record is saved in line 95. Let me know your thoughts.

File: Tests/reviews.service.test.js (hunk 3 above)
Comment (bshresthap) [reply Tests/reviews.service.test.js:84]: Made note of it. Planning to take care of it for next PR.

File: database-mongoose/database.js (hunk 4 above)
Comment (bshresthap) [reply database-mongoose/database.js]: Made note of it. Planning to take care of it for next PR.

File: database-mongoose/reviews.service.js (hunk 5 above)
Comment (bshresthap) [reply database-mongoose/reviews.service.js]: Valid point. Thanks Alek. Made note of it. Planning to take care of it for next PR.

File: database-mongoose/reviews.service.js (hunk 6 above)
Comment (bshresthap) [reply database-mongoose/reviews.service.js]: Made note of it. Planning to take care of it for next PR.

File: database-mongoose/reviews.service.js (hunk 7 above)
Comment (bshresthap) [reply database-mongoose/reviews.service.js]: Made note of it. Planning to take care of it for next PR. I  am changing the code to Insert many.

File: database-mongoose/reviews.service.js (hunk 9 above)
Comment (bshresthap) [reply database-mongoose/reviews.service.js:101]: Great point Alek. Made note of it. Planning to take care of it for next PR.

File: database-mongoose/reviews.service.js (hunk 10 above)
Comment (bshresthap) [reply database-mongoose/reviews.service.js:142]: Made note of it. Planning to take care of it for next PR.

File: database-mongoose/seeder.js (hunk 11 above)
Comment (bshresthap) [reply database-mongoose/seeder.js:80]: Yes, will be cleaning up comments for next PR.

File: server/server.js (hunk 12 above)
Comment (bshresthap) [reply server/server.js:54]: Good catch Alek. Will take care of it when I am working on unit tests - for next PR.

**Comment by bshresthap** [comment]
@aleksebastian Really appreciate your feedback and found it valuable. Please review my comments/response. One outstanding conversation might be the related to Tests/reviews.model.test.js. Once all conversations are resolved, we can plan for code merge.

File: Tests/reviews.model.test.js (hunk 1 above)
Comment (aleksebastian) [reply Tests/reviews.model.test.js:15]: Excellent, did not notice you are using Jest here. Looks super similar to mocha.

File: Tests/reviews.model.test.js (hunk 2 above)
Comment (aleksebastian) [reply Tests/reviews.model.test.js:99]: I see it now. Makes sense!

File: database-mongoose/reviews.service.js (hunk 7 above)
Comment (aleksebastian) [reply database-mongoose/reviews.service.js]: 👍
